ompi/mca/coll/han/coll_han_reduce.c

```diff
@@ -176,7 +176,7 @@ mca_coll_han_reduce_intra(const void *sbuf,
             t->sbuf = (char *) t->sbuf + extent * t->seg_count;
         }
 
-        if (up_rank == root_up_rank) {
+        if (up_rank == root_up_rank && NULL != t->rbuf) {
```

Comment:
Hmmm should this be `if (root_low_rank == low_rank)` instead?

Reading L142-L148, only the root and node leaders should have valid rbuf.

Reply:
I'd say no, because the root has a fully sized rbuf, while non-root node leaders only have a temporary buffer that fits two segments (L122, L146). But what we *could* do is `if (up_rank == root_up_rank && low_rank == root_low_rank)`. Or `if (w_rank == root)`; should also be equivalent. These might indeed be more in tune with the rest of the code (opinions welcome).

Let me take another look at the code, perhaps one or two touchups in related spots would be a good idea, to make it all a bit clearer.

Reply:
`up_rank == root_up_rank` is true only for the root of the original reduce call, so the only process that has a user provided output buffer. Thus, it makes sense to move to base pointer as we advance in the segments.

For all leaders, aka. processes that become root for the local reductions, with the exception of the original root we only allocated a temporary buffer with size `2*extent*seg_count`, so there is no room to move the pointer around (more than the double buffering in `mca_coll_han_reduce_t1_task`).

Reply:
> `up_rank == root_up_rank` is true only for the root of the original reduce call

-> For the global root, and for the ranks in the same node as it, right?